rezervo/integrations/fsc/sessions.py

```diff
@@ -0,0 +1,125 @@
+from typing import List, Optional
+from uuid import UUID
+
+import pydantic
+import requests
+
+from rezervo import models
+from rezervo.database.database import SessionLocal
+from rezervo.errors import AuthenticationError
+from rezervo.integrations.fsc.auth import authenticate_session
+from rezervo.integrations.fsc.booking import booking_url
+from rezervo.integrations.fsc.schedule import fetch_fsc_schedule
+from rezervo.integrations.fsc.schema import (
+    BookingData,
+    BookingsResponse,
+    FscClass,
+    rezervo_class_from_fsc_class,
+    session_state_from_fsc,
+)
+from rezervo.models import SessionState
+from rezervo.schemas.config.user import (
+    IntegrationConfig,
+    IntegrationIdentifier,
+    IntegrationUser,
+    get_integration_config_from_integration_user,
+)
+from rezervo.schemas.schedule import UserSession
+from rezervo.utils.logging_utils import err
+
+
+def fetch_fsc_sessions(user_id: Optional[UUID] = None) -> dict[UUID, list[UserSession]]:
+    fsc_schedule = fetch_fsc_schedule()
+    with SessionLocal() as db:
+        db_fsc_users_query = db.query(models.IntegrationUser).filter(
+            models.IntegrationUser.integration == IntegrationIdentifier.FSC
+        )
+        if user_id is not None:
+            db_fsc_users_query = db_fsc_users_query.filter(
+                models.IntegrationUser.user_id == user_id
+            )
+        db_fsc_users = db_fsc_users_query.all()
+        sessions: dict[UUID, list[UserSession]] = {}
+        for db_fsc_user in db_fsc_users:
+            fsc_user: IntegrationUser = IntegrationUser.from_orm(db_fsc_user)
+            auth_session = authenticate_session(fsc_user.username, fsc_user.password)
+            if isinstance(auth_session, AuthenticationError):
+                err.log(
+                    f"Authentication failed for '{fsc_user.username}', abort user sessions pull!"
+                )
+                continue
+            try:
+                res = auth_session.get(booking_url(auth_session))
+            except requests.exceptions.RequestException as e:
+                err.log(
+                    f"Failed to retrieve sessions for '{fsc_user.username}'",
+                    e,
+                )
+                continue
+            bookings_response: BookingsResponse = res.json()
+            fsc_sessions = []
+            if "data" in bookings_response:
+                for s in bookings_response["data"]:
+                    fsc_sessions.append(pydantic.parse_obj_as(BookingData, s))
+            past_and_imminent_sessions = [
+                UserSession(
+                    integration=IntegrationIdentifier.FSC,
+                    class_id=s.groupActivity.id,
+                    user_id=fsc_user.user_id,
+                    status=session_state_from_fsc(s.type),
+                    class_data=rezervo_class_from_fsc_class(
+                        next(
+                            (c for c in fsc_schedule if c.id == s.groupActivity.id),
```

Comment:
Could make it even more fancy with [`filter`](https://docs.python.org/3/library/functions.html#filter)